lib/windows/app/actions/deviceActions.js

```diff
@@ -0,0 +1,313 @@
+
+        let message = `Error while probing usb device at bus.address ${usbAddr}: ${error.message}. `;
+        if (process.platform === 'linux') {
+            message += 'Please check your udev rules concerning permissions for USB devices, see ' +
+                'https://github.com/NordicSemiconductor/nrf-udev';
+        } else if (process.platform === 'win32') {
+            message += 'Please check that a libusb-compatible kernel driver is bound to this device.';
+        }
+
+        logger.error(message);
+    } else {
+        logger.error(`Error while probing devices: ${error.message}`);
+    }
+}
+
+/**
+ * Deselects the currently selected device.
+ *
+ * @returns {Object} Action object that can be passed to redux dispatch.
+ */
+export function deselectDevice() {
+    return {
+        type: DEVICE_DESELECTED,
+    };
+}
+
+/**
+ * Starts watching for devices with the given traits. See the nrf-device-lister
+ * library for available traits. Whenever devices are attached/detached, this
+ * will dispatch DEVICES_DETECTED with a complete list of attached devices.
+ *
+ * @returns {function(*)} Function that can be passed to redux dispatch.
+ */
+export function startWatchingDevices() {
+    const config = getAppConfig();
+
+    return (dispatch, getState) => {
+        deviceLister = new DeviceLister(config.selectorTraits);
+
+        deviceLister.on('conflated', devices => {
+            const state = getState();
+            if (state.core.device.selectedSerialNumber !== null &&
+                !devices.has(state.core.device.selectedSerialNumber)) {
+                dispatch(deselectDevice());
+            }
+
+            dispatch(devicesDetectedAction(Array.from(devices.values())));
+        });
+        deviceLister.on('error', logDeviceListerError);
+        deviceLister.start();
+    };
+}
+
+/**
+ * Stops watching for devices.
+ *
+ * @returns {function(*)} Function that can be passed to redux dispatch.
+ */
+export function stopWatchingDevices() {
+    return () => {
+        deviceLister.removeAllListeners('conflated');
+        deviceLister.removeAllListeners('error');
+        deviceLister.stop();
+    };
+}
+
+/**
+ * Asks the user to provide input during device setup. If a list of choices are
+ * given, and the user selects one of them, then then promise will resolve with
+ * the selected value. If no choices are given, and the user confirms, then the
+ * promise will just resolve with true. Will reject if the user cancels.
+ *
+ * @param {function} dispatch The redux dispatch function.
+ * @param {String} message The message to display to the user.
+ * @param {Array<String>} [choices] The choices to display to the user (optional).
+ * @returns {Promise<String>} Promise that resolves with the user input.
+ */
+function getDeviceSetupUserInput(dispatch, message, choices = undefined) {
+    return new Promise((resolve, reject) => {
+        deviceSetupCallback = choice => {
```

Comment:
This unconditional assignment looks like it's prone to race conditions. The code around https://github.com/NordicSemiconductor/pc-nrfconnect-core/pull/191/files#diff-c98e84aebf23c3418e027d39d17f5b6dR306 will execute the last callback set.

Reply:
Why not pass `deviceSetupCallback` as an argument to `deviceSetupInputRequiredAction` instead? That'd keep the callback scope in check.

Reply:
You cannot and should not store a function in an immutable state record. Also only local functions executed in simple state machine-like manner effect this assignment, there is nothing that can race here. I wouldn't worry about it.